reducing the image size

seffenberg-naspers opened the pull request:
- the default build will now use 200-300MB less disk space than the original
  one
- if the build-arg CLEAN is set to 1 then also all build dependencies
  and the ./test and ./src directories are removed so that the image is
  ~ 480 MB big instead of 1.8GB
  
  --build-arg CLEAN=1

TODO: discuss if the default of CLEAN should be 1 instead of 0

savar committed:
reducing the image size

* the default build will now 200-300MB less disk space than the original
  one
* if the build-arg CLEAN is set to 1 then also all build dependencies
  and the ./test and ./src directories are removed so that the image is
  ~ 480 MB big instead of 1.8GB

  --build-arg CLEAN=1

TODO: discuss if the default of CLEAN should be 1 instead of 0

AkeemMcLennon commented:
Sorry for the late reply! I'll take a look at this later tonight and make sure everything still works as expected. I might also take the opportunity to merge upstream phantomjs / ghost changes.

ghost commented:
Any updates on this?

AkeemMcLennon commented on Dockerfile:
```diff
@@ -2,20 +2,37 @@ FROM ubuntu:trusty
+                            libpng-dev libjpeg-dev \
+                            git-core libicu52 \
+                            curl \
+      && curl -L https://github.com/AkeemMcLennon/phantomjs/archive/2.0.0.tar.gz | tar --strip-components=1 -xzf - \
```
@seffenberg-naspers Just curious, was there a specific reason you moved this all into a single RUN command? 

Although I understand what you're doing, this change makes the diff a little less clear and doesn't make use of any build caching.

seffenberg-naspers replied:
The problem is, that each RUN command will create an intermediate image. If this is bigger than the next one (because the next command is a `apt-get clean` or stuff like this) then this next image won't shrink the image afterwards. So doing this in multiple RUN commands means, the resulting image will have huge intermediate images you have to download as well. Thats what I want to avoid (so not downloading 2GB but only ~450MB).

AkeemMcLennon replied:
I see. Makes sense.

AkeemMcLennon replied:
If you're downloading the repository as an archive, you'll need to download from master instead of the 2.0.0 release. Otherwise, the zip file won't contain the ghostdriver patch.

seffenberg-naspers replied:
so the 2.0.0 release (tag) or the 2.0 release (branch) aren't okay? I can change that to master.zip.

seffenberg-naspers replied:
I can easily change that if you say it is necessary! Only ping me @AkeemMcLennon

AkeemMcLennon replied:
@seffenberg-naspers Yeah, you'll need to change this for selenium to work. I've forked a separate version of phantomjs since they haven't merged the upstream ghostdriver in two years. Eventually I'll rebase against the newest versions while reapplying the patch, but for now just point to the master branch of this repository.

seffenberg-naspers replied:
@AkeemMcLennon done.. can you please re-check if it works well?

AkeemMcLennon commented on Dockerfile:
```diff
@@ -2,20 +2,37 @@ FROM ubuntu:trusty
+      && curl -L https://github.com/AkeemMcLennon/phantomjs/archive/2.0.0.tar.gz | tar --strip-components=1 -xzf - \
+      && ./build.sh --confirm \
+      && ln -s /usr/local/phantomjs/bin/phantomjs /usr/bin/phantomjs \
+      && ( [ "$CLEAN" -ne 0 ] && ( \
```
Thank you for the clean up! Could you make lines 24-32 a separate RUN command?

seffenberg-naspers replied:
see my comment from before.. it is the same reason ;)

AkeemMcLennon commented on Dockerfile:
```diff
@@ -2,20 +2,37 @@ FROM ubuntu:trusty
+
+COPY join-hub.sh /usr/local/phantomjs/
+
+ENTRYPOINT [ "sh", "-c", "/usr/local/phantomjs/join-hub.sh" ]
```
Good catch! The additional argument is a good idea.

seffenberg-naspers replied:
Thanks :-), the only thing is, I would need an image with CLEAN=1 which I can pull, so if you could either do this by default (setting CLEAN=1 in the beginning) and the next tagged image is cleaned or I need another image next to the "normal" one which maybe is named like the "normal" one but with a suffix `-cleaned` or something like that

AkeemMcLennon replied:
It appears that the "-c" argument also enforces file permission checks.  Unless you address the file permissions in the Dockerfile , you'll get the following (since git would reset these permissions on checkout).

`sh: 1: /usr/local/phantomjs/join-hub.sh: Permission denied`

seffenberg-naspers replied:
thanks for that.. I changed the file mode (see below)

savar committed:
file needs to be executable

savar committed:
point to master branch and not to the version 2.0.0

AkeemMcLennon commented:
Looks good! Let's just go ahead and make CLEAN the default. After that's done, I'll go ahead and merge.

savar committed:
the default built should CLEAN

the default built should clean out unnecessary files so that the created
image is as small as possible

ghost commented:
done.. `CLEAN=1` is now in the Dockerfile
